lewis/adapters/modbus.py

```diff
@@ -0,0 +1,569 @@
+# -*- coding: utf-8 -*-
+# *********************************************************************
+# lewis - a library for creating hardware device simulators
+# Copyright (C) 2016 European Spallation Source ERIC
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# *********************************************************************
+# References Used:
+#   http://www.modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf
+#   http://www.modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf
+#   https://github.com/sourceperl/pyModbusTCP
+#   https://github.com/bashwork/pymodbus
+# *********************************************************************
+
+
+from __future__ import division
+
+import socket
+import asyncore
+import struct
+
+from copy import deepcopy
+from math import ceil
+from argparse import ArgumentParser
+
+from lewis.adapters import Adapter
+
+
+class ModbusDataBank(object):
+    def __init__(self, config):
+        """
+        Preliminary DataBank implementation for Modbus.
+
+        This is a very generic implementation of a databank for Modbus. It's meant to set the
+        groundwork for future implementations. The create_basic method should be used for
+        construction, rather than this initializer. The signature of this __init__ method is
+        subject to change.
+
+        .. sourcecode:: Python
+
+            di = ModbusDataBank.create_basic(False, 0x1000, 0x1FFF)
+
+        :param config: dict containing configuration
+        """
+        self._data = config['data']
+        self._start_addr = config['start_addr']
+
+    @classmethod
+    def create_basic(cls, default_value=0, start_addr=0x0000, last_addr=0xFFFF):
```

Comment:
That would work as well... But I imagine we will eventually wind up with many tiny classes that are almost identical, for the various scenarios. 

Inheritance makes most sense when you have a common interface but differing implementations. In this case, I think both interface and implementation can/will always be the same. But the initialization will differ depending on what kind of DataBank you want. Consider:

```python
@classmethod
def create_remapped(cls, databank, new_start_addr):
    return cls({'start_addr': new_start_addr, 'data': databank._data})
```
```python
di = ModbusDataBank.create_basic(False, 0x1000, 0x1FFF)
co = ModbusDataBank.create_remapped(di, 0x2000)
```
Now `co` addresses 0x2000 - 0x2FFF map to the same memory as `di` address 0x1000 - 0x1FFF.

You could also have various types of conversion between values and address spaces of bit-based DataBanks and word-based DataBanks. DataBanks that bind to attributes / functions would have `create_bound` or similar. The internal implementation of ModbusDataBank (including `get` and `set`) would of course have to be extended significantly to support all this. 

One advantage of doing it this way is that it emphasizes that a DataBank is always fundamentally the same type of object; all that changes is how it is set up.

But I'm not very attached to this pattern, we could certainly achieve the same with inheritance. Just felt like the natural way to go... I think I saw something similar used in pymodbus somewhere and thought "ah, we could do even more with this".

You're right it's very different from what we've done in other places. I'm happy to change it if you think it's too far from our usual MO.